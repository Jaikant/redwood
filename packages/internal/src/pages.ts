import path from 'path'
import fs from 'fs'

export type PagesDependency = {
  const: string
  path: string
  importStatement: string
}

/**
 * Recursively process the pages directory and return information useful for
 * automated imports.
 */
export const processPagesDir = (
  webPagesDir: string,
  prefix: Array<string> = []
): Array<PagesDependency> => {
  const deps: Array<PagesDependency> = []
  const entries = fs.readdirSync(webPagesDir, { withFileTypes: true })

  // Iterate over a dir's entries, recursing as necessary into
  // subdirectories.
  entries.forEach((entry) => {
    if (entry.isDirectory()) {
      // Actual JS files reside in a directory of the same name, so let's
      // construct the filename of the actual Page file.
      const testFile = path.join(webPagesDir, entry.name, entry.name + '.js')

      if (fs.existsSync(testFile)) {
        // If the Page exists, then construct the dependency object and push it
        // onto the deps array.
        const basename = path.posix.basename(entry.name, '.js')
        const importName = prefix.join() + basename
        // `src/pages/<PageName>`
        const importFile = ['src', 'pages', ...prefix, basename].join('/')
        deps.push({
          const: importName,
          path: path.join(webPagesDir, entry.name),
          importStatement: `const ${importName} = { name: '${importName}', loader: () => import('${importFile}') }`,
        })
      } else {
        // If the Page doesn't exist then we are in a directory of Page
        // directories, so let's recurse into it and do the whole thing over
        // again.
        const newPrefix = [...prefix, entry.name]
        deps.push(
          ...processPagesDir(path.join(webPagesDir, entry.name), newPrefix)
        )
      }
    }
  })
  return deps
}
