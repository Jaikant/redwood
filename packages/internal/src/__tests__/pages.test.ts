import MockProject, { DEFAULT_PROJECT_PATH } from '@redwoodjs/test-mocks'

import { processPagesDir } from '../pages'

describe('pages', () => {
  const mockrw = new MockProject()

  beforeEach(() => {
    mockrw.mock()
  })

  afterAll(() => {
    mockrw.restore()
  })

  describe('processPagesDir', () => {
    it('finds the pages and returns the correct things', () => {
      mockrw.mergePaths(() => ({
        web: {
          src: {
            pages: {
              HomePage: { 'HomePage.js': '' },
              AboutPage: { 'AboutPage.js': '' },
              Admin: { UsersPage: { 'UsersPage.js': '' } },
            },
          },
        },
      }))

      const pages = processPagesDir(`${DEFAULT_PROJECT_PATH}/web/src/pages`)

      mockrw.restore()
      expect(pages).toMatchSnapshot()
    })
  })
})
