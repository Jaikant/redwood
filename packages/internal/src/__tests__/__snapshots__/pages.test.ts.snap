// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages processPagesDir finds the pages and returns the correct things 1`] = `
Array [
  Object {
    "const": "AboutPage",
    "importStatement": "const AboutPage = { name: 'AboutPage', loader: () => import('src/pages/AboutPage') }",
    "path": "/mock/project/web/src/pages/AboutPage",
  },
  Object {
    "const": "AdminUsersPage",
    "importStatement": "const AdminUsersPage = { name: 'AdminUsersPage', loader: () => import('src/pages/Admin/UsersPage') }",
    "path": "/mock/project/web/src/pages/Admin/UsersPage",
  },
  Object {
    "const": "HomePage",
    "importStatement": "const HomePage = { name: 'HomePage', loader: () => import('src/pages/HomePage') }",
    "path": "/mock/project/web/src/pages/HomePage",
  },
]
`;
